### PR TITLE
DBACKUP-3

### PR DESCRIPTION
**Pull request by carapai**
This pull request is based on issue
https://issues.openmrs.org/browse/DBACKUP-3
Which Improves performance of the data extraction on very large tables especially obs

### CODE REVIEW

**Comment by carapai** [comment]
@dkayiwa 
I have deleted the other pull request because of many commits.
Please review this one and advice if there is anything I am missing.

**Comment by dkayiwa** [comment]
@carapai did you actually test this out live?

File: api/src/main/java/org/openmrs/module/databasebackup/util/DbDump.java [hunk 1]
```diff
@@ -183,42 +183,83 @@ public static void dumpDB(Properties props, boolean showProgress, Class showProg
...
-                        outputValue = escape(outputValue);
-                        result.write( "'"+outputValue+"'" );
+                        result.write("),");
                     }
                 }
-                result.write(");\n");
+                rs.close();
+                stmt.close();
+
+                if (offset >= count || count <= max) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/databasebackup/util/DbDump.java:251]: You could as well just break out of this loop.

**Comment by carapai** [comment]
Yes I tested on database with 4 million obs and worked.
@ssmusoke can also confirm this since He was also doing some tests

File: api/src/main/java/org/openmrs/module/databasebackup/util/DbDump.java (hunk 1 above)
Comment (carapai) [reply api/src/main/java/org/openmrs/module/databasebackup/util/DbDump.java:251]: The first loop breaks if all the rows have been processed, the offset is greater than total number of rows in the table. I am not getting exactly why you say just break out of the loop

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/databasebackup/util/DbDump.java:251]: I mean just using the break statement.

**Comment by dkayiwa** [comment]
@carapai so are you doing batch inserts of 10000?

**Comment by carapai** [comment]
Yes

**Comment by dkayiwa** [comment]
Hopefully that will stay within max_allowed_packet

**Comment by dkayiwa** [comment]
Uncle @ssmusoke did you test this out?

**Comment by dkayiwa** [comment]
@carapai around how many times faster than before is this?

**Comment by carapai** [comment]
First of the other one would not even backup obs if it is more than 2000000 rows
I have not calculated the difference in time. But based on what we had in the field, if your database has many obs, some time it could take even 4 hours, but this one is similar to default mysqldump.

**Comment by dkayiwa** [comment]
It would be great to do some rough statistical comparisons.

**Comment by carapai** [comment]
OK, but one fact is that the other one does not even backup. What we had in the field is not even 1000000 obs. Never the less i will do some comparisons and get back to you

**Comment by dkayiwa** [comment]
FWIW, the performance comparisons are not a precondition for merging. They are simply for us to know how better we have gone. 😊 

All because your pull request is dealing with two things instead of one.
The performance changes should have been a pull request of their own. So next time .... 👶

**Comment by ssmusoke** [comment]
@dkayiwa Just run a test on a large database and it backups up well plus restore - this is good to go for merging

**Pull request merged**